### PR TITLE
Fix SelectExpandNode for navigation properties on derived/complex types

### PR DESCRIPTION
**Pull request by mikepizzo**
### Issues

PR #1286 fixed a number of issues related to how non-selected expanded navigation properties were represented in the ContextUrl. However, in the new code that attempts to determine if the entire subtree is selected, it looks only at direct children to see if they are navigation properties and not deeper.  

For example, prior to this fix the following URL would return a SelectExpandNode with SelectionType=PartialSubtree.

             Employees?$expand=ns.Manager/DirectReports

After the PR, this is correctly returned as EntireSubtree

### Description

The primary purpose of this PR is to set whether the SelectNode is entiresubtree based on a deep traverse of child paths, rather than only looking at direct children.

Other changes, in addition to new tests, include some test clean-up as well as code optimizations around deferring creation of hashsets until required.

### Checklist (Uncheck if it is not completed)

- [x ] *Test cases added*
- [x ] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: src/Microsoft.OData.Core/SelectedPropertiesNode.cs [hunk 1]
```diff
@@ -117,10 +117,23 @@ private void Parse(string selectClause)
                 this.ParsePathSegment(segments, 0);
             }
 
-            if ( (this.selectedProperties == null || this.selectedProperties.Count == 0)
-                && this.children.Values.All( n => n.selectionType == SelectionType.EntireSubtree))
+            DetermineSelectionType(this);
+        }
+
+        private static void DetermineSelectionType(SelectedPropertiesNode node)
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/SelectedPropertiesNode.cs]: >DetermineSelectionType [](start = 28, length = 22)

I'd suppose to move the private static method end of this class.

Comment (xuzhg) [inline src/Microsoft.OData.Core/SelectedPropertiesNode.cs]: >SelectedPropertiesNode node [](start = 51, length = 27)

It seems we also can use instance method, not the static. But, either way is ok for me. Only to call this static method using "this" a little bit confused me.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/SelectedPropertiesNodeTests.cs [hunk 2]
```diff
@@ -63,8 +63,16 @@ public SelectedPropertiesNodeTests()
             this.defaultContainer.AddEntitySet("Cities", cityType);
             this.defaultContainer.AddEntitySet("Districts", districtType);
 
+            EdmComplexType airportType = new EdmComplexType("TestModel", "Airport");
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/SelectedPropertiesNodeTests.cs]: >airportType [](start = 27, length = 11)

should we add "airportType" into edm model? maybe i miss that.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/SelectedPropertiesNodeTests.cs [hunk 3]
```diff
@@ -442,7 +462,51 @@ public void ExpandTokenForCollectionOfNavigationPropertyShouldHaveEntireSubtree(
...
+        }
+
+        [Fact]
+        public void ExpandTokenForNavigationPropertyOnDerivedTypeShouldHaveEntireSubtree()
+        {
+            SelectedPropertiesNode.Create("TestModel.Metropolis/MetropolisNavigation()", this.cityType, this.edmModel).Should().HaveEntireSubtree();
+        }
+
+        [Fact]
+        public void ExpandTokenForNavigationPropertyDerivedOnComplexShouldHaveEntireSubtree()
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/SelectedPropertiesNodeTests.cs]: >DerivedOn [](start = 52, length = 9)

Maybe "....OnDerivedComplex...."

File: test/FunctionalTests/Microsoft.OData.Core.Tests/SelectedPropertiesNodeTests.cs [hunk 4]
```diff
@@ -442,7 +462,51 @@ public void ExpandTokenForCollectionOfNavigationPropertyShouldHaveEntireSubtree(
...
+        }
+
+        [Fact]
+        public void ExpandTokenForNavigationPropertyOnComplexShouldHaveEntireSubtree()
+        {
+            SelectedPropertiesNode.Create("NearestAirport/City()", this.metropolisType, this.edmModel).Should().HaveEntireSubtree();
+        }
+
+        [Fact]
+        public void ExpandTokenForNavigationPropertyDerivedOnComplexWithSelectShouldHavePartialSubtree()
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/SelectedPropertiesNodeTests.cs]: >DerivedOnComplex [](start = 52, length = 16)

Maybe "...OnDerivedComplex..."

File: src/Microsoft.OData.Core/SelectedPropertiesNode.cs [hunk 5]
```diff
@@ -791,16 +790,17 @@ private bool IsOperationSelectedAtThisLevel(IEdmOperation operation, bool mustBe
...
             }
 
-            if (this.selectionType == SelectionType.EntireSubtree)
+            if (this.selectionType == SelectionType.Empty || this.selectedProperties == null )
             {
-                return true;
+                return false;
             }
 
+
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/SelectedPropertiesNode.cs]: remove the un-used new line

**Comment by xuzhg** [comment]
The change looks good to me,  but left some small comments. Thanks

File: test/FunctionalTests/Microsoft.OData.Core.Tests/SelectedPropertiesNodeTests.cs (hunk 2 above)
Comment (mikepizzo) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/SelectedPropertiesNodeTests.cs]: This and regionalAirportType seemed to be implicitly added to the model, but I made them explicit.

---
In reply to: [242373747](https://github.com/OData/odata.net/pull/1359#discussion_r242373747) [](ancestors = 242373747)

File: src/Microsoft.OData.Core/SelectedPropertiesNode.cs (hunk 1 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/SelectedPropertiesNode.cs]: Yes, I like that better.

---
In reply to: [242370658](https://github.com/OData/odata.net/pull/1359#discussion_r242370658) [](ancestors = 242370658)

**Pull request merged**